File: cos/collector/collector.py
# -*- coding:utf-8 -*-
# Copyright 2024 coScene
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
import logging.config
import textwrap
from datetime import datetime
from typing import List
from multiprocessing import Queue

from pydantic import BaseModel, Field
from strictyaml import load, YAMLError

from cos.constant import RECORD_DIR_PATH
from cos.core.api import ApiClient, ApiClientState
from cos.core.models import FileInfo, RecordCache
from cos.utils import is_image
from .codes import EventCodeManager
from ..core import request_hook
from ..core.exceptions import Unauthorized, RecordNotFound
from ..utils.files import can_read_path

_log = logging.getLogger(__name__)


class DeviceConfig(BaseModel):
    extra_files: List[str] = Field(default_factory=list)


class CollectorConfig(BaseModel):
    delete_after_upload: bool = True
    delete_after_interval_in_hours: int = 48
    scan_interval_in_secs: int = 60
    skip_check_same_file: bool = False


class Collector:
    def __init__(
        self,
        conf: CollectorConfig,
        device_conf: DeviceConfig,
        api_client: ApiClient,
        code_manager: EventCodeManager,
    ):
        self.conf = conf
        self.device_conf = device_conf
        self.api = api_client
        self.code_mgr = code_manager
        self.device = self.api.state.load_state().device

    def _upload_record_thumbnail(self, record_name, rec_cache: RecordCache):
        if not record_name:
            return
        for f in rec_cache.file_infos:
            if is_image(f.filename):
                upload_url = self.api.generate_record_thumbnail_upload_url(record_name)
                if upload_url:
                    self.api.upload_file(str(f.filepath), upload_url)
                    return

    def _upload_finish_flag_file(self, record_name, rec_cache: RecordCache) -> bool:
        _finish_file = rec_cache.base_dir_path / "finish.flag"
        if not _finish_file.exists():
            _finish_file.touch()
            with _finish_file.open("w", encoding="utf8") as fp:
                fp.write(json.dumps(rec_cache.files, indent=2))

        file_info = FileInfo(
            filepath=str(_finish_file.absolute()),
            filename=_finish_file.name,
        ).complete(inplace=True, skip_sha256=True)
        return self.api.resumable_upload_files(
            record_name=record_name,
            file_infos=[file_info],
            part_state_path=str(rec_cache.base_dir_path.absolute()),
            skip_check_same_file=self.conf.skip_check_same_file,
        )

    def __get_record_title(self, rec_cache: RecordCache):
        title = rec_cache.record.get("title", None)
        if title:
            return title
        if rec_cache.task.get("title", None):
            return rec_cache.task.get("title")

        code = rec_cache.event_code
        msg = self.code_mgr.get_message(code)
        trigger_time = datetime.fromtimestamp(rec_cache.timestamp / 1000).isoformat()
        return f'{msg or "未知错误"} ({code}) @ {trigger_time}'

    def __make_record_description(self, title, rec_cache: RecordCache):
        description = rec_cache.record.get("description", None)
        if description:
            return description

        result = textwrap.dedent(
            f"""\
            ### {title} \n
            the record is triggered @ {rec_cache.timestamp} \n
            the files are from {rec_cache.base_dir_path} \n
            on robot: {self.device.get("serial_number", "")} \n
        """
        )
        if self.device:
            for label in self.device.get("labels", []):
                result += "\n" + label.get("displayName", "")
        return result

    def _load_device_extra_infos(self):
        infos = {}

        for p in self.device_conf.extra_files:
            if not can_read_path(p):
                continue

            if not (p.endswith(".yml") or p.endswith(".yaml")):
                continue

            with open(p, "r", encoding="utf8") as y:
                try:
                    robot_yaml = load(y.read()).data
                except YAMLError as exc:
                    raise ValueError(f"Failed to load device extra yaml: {exc}")
            # merge robot yaml into infos dict
            infos.update(robot_yaml)
        return infos

    def _create_record(self, rec_cache: RecordCache):
        record_title = self.__get_record_title(rec_cache)
        record = self.api.create_or_get_record(
            file_infos=rec_cache.file_infos,
            title=record_title,
            description=self.__make_record_description(record_title, rec_cache),
            labels=rec_cache.labels,
            device_name=self.device.get("name"),
            record_name=rec_cache.record.get("name") if rec_cache.record else None,
            reserve_file_infos=True,
            rules=rec_cache.record.get("rules", []),
        )
        return record

    def _create_record_related_resources(self, rec_cache: RecordCache):
        record_name = rec_cache.record.get("name", "")
        record_title = rec_cache.record.get("title", "")
        if not record_name:
            return

        for moment in rec_cache.moments:
            if not moment.name:
                # 当前部分组件使用的是毫秒级别的时间戳，所以这里需要转换
                if moment.timestamp > 1_000_000_000_000:
                    moment.timestamp /= 1000

                obtain_event_res = self.api.obtain_event(
                    record_name=record_name,
                    display_name=moment.title if moment.title else record_title,
                    description=moment.description if moment.description else record_title,
                    customized_fields=moment.metadata,
                    trigger_time=moment.timestamp,
                    duration=moment.duration,
                    rule_id=moment.rule_id,
                    device_name=self.device["name"],
                )
                moment.name = obtain_event_res.get("event", {}).get("name")
                moment.is_new = obtain_event_res.get("isNew", False)
                rec_cache.save_state()

            if moment.name and not moment.event:
                has_sent = moment.event.get("sent", False)
                if not has_sent:
                    device_extra_infos = self._load_device_extra_infos()

                    self.api.trigger_device_event(
                        moment=moment,
                        record_name=record_name,
                        device_name=self.device.get("name"),
                        device_extra_info=device_extra_infos,
                        event_code=moment.code,
                        rule_id=moment.rule_id,
                    )
                    moment.event["sent"] = True
                    rec_cache.save_state()
                    _log.info(f"==> Triggered event: {moment.name}")

            if not moment.is_new:
                continue

            if not moment.task:
                continue

            if not moment.task.name:
                upserted_task = self.api.upsert_task(
                    record_name=record_name,
                    event_name=moment.name,
                    title=moment.title if moment.title else record_title,
                    description=moment.description if moment.description else record_title,
                    assignee=moment.task.assignee,
                )

                moment.task.name = upserted_task.get("name")
                rec_cache.save_state()
                if upserted_task.get("name") and moment.task.sync_task:
                    self.api.sync_task(upserted_task.get("name"))

        if rec_cache.diagnosis_task and not rec_cache.diagnosis_task.get("name"):
            created_diagnosis_task = self.api.create_diagnosis_task(
                title=record_title,
                description=self.__make_record_description(record_title, rec_cache),
                device=self.device.get("name"),
                rule_id=rec_cache.diagnosis_task.get("rule_id"),
                rule_name=rec_cache.diagnosis_task.get("rule_name"),
                trigger_time=rec_cache.diagnosis_task.get("trigger_time"),
                start_time=rec_cache.diagnosis_task.get("start_time"),
                end_time=rec_cache.diagnosis_task.get("end_time"),
                record_name=record_name,
            )
            rec_cache.diagnosis_task["name"] = created_diagnosis_task.get("name")
            rec_cache.save_state()

    def handle_record(self, rec_cache: RecordCache):
        _log.debug(f"==> Checking record: {rec_cache.key}")
        # setup project name
        if rec_cache.project_name:
            self.api.project_name = rec_cache.project_name
        else:
            self.api.project_name = None

        # 0. 如果 skipped，直接退出
        if rec_cache.skipped:
            _log.debug(f"==> Record previously skipped: {rec_cache.key}")

        # 1. 如果还没创建过 record, 检查是否超过了 code limit
        elif not rec_cache.record.get("name") and rec_cache.event_code and self.code_mgr.is_over_limit(rec_cache.event_code):
            _log.warning(f"==> Reached code limit {rec_cache.event_code}, skip handle: {rec_cache.key}")
            task_name = rec_cache.task.get("name", "")
            if task_name:
                self.api.update_task_state(task_name, "SUCCEEDED")

            # 把跳过标记写回到json
            rec_cache.skipped = True
            rec_cache.save_state()
            rec_cache.delete_cache_dir(self.conf.delete_after_interval_in_hours)

        else:
            # 如果没有被 collected (文件未找齐).
            if not rec_cache.record.get("name"):
                # 2. 收集文件
                rec_cache.file_infos = [
                    FileInfo(
                        filepath=f.filepath.resolve().absolute(),
                        filename=f.filename,
                    ).complete(inplace=True, skip_sha256=True)
                    for f in rec_cache.file_infos
                    if f.filepath.is_file() and f.filename != "finish.flag" and can_read_path(str(f.filepath.absolute()))
                ]

                # 3. 创建 record 和 event
                rec_cache.record = self._create_record(rec_cache)
                # 为防止断连后重新建立记录，需要立即写回json
                rec_cache.save_state()
                self.code_mgr.hit(rec_cache.event_code)

                task_name = rec_cache.task.get("name", "")
                if task_name:
                    self.api.put_task_tags(task_name, {"recordName": rec_cache.record.get("name", "")})

                # 4. 上传缩略图
                self._upload_record_thumbnail(rec_cache.record.get("name"), rec_cache)

            # 检查关联的资源是否已经上传
            self._create_record_related_resources(rec_cache)

            # 找齐文件后，如果还没有 uploaded (上传完毕).
            if not rec_cache.uploaded:
                _uploaded_files = rec_cache.uploaded_filepaths

                # 5. 等待上传文件清单
                file_infos = []
                for f in rec_cache.file_infos:
                    if can_read_path(str(f.filepath.absolute())):
                        file_infos.append(f)
                    else:
                        _log.warning(f"{f.filepath} can not access, skip!")
                sorted_files: List[FileInfo] = sorted(file_infos, key=lambda f: f.size)

                all_completed = True
                for file_info in sorted_files:
                    _abs_path = str(file_info.filepath.absolute())

                    if not can_read_path(_abs_path):
                        _log.warning(f"{file_info.filepath} can not access, skip!")
                        continue

                    if _abs_path in _uploaded_files:
                        _log.debug(f"{file_info.filepath} already uploaded, skip!")
                        continue

                    if not rec_cache.state_path.absolute().exists():
                        _log.warning(f"{rec_cache.state_path.absolute()} not exist, skip!")
                        return

                    _rc = RecordCache.load_state_from_disk(rec_cache.state_path.absolute())
                    if _rc.skipped:
                        return

                    _completed = self.api.resumable_upload_files(
                        record_name=rec_cache.record["name"],
                        file_infos=[file_info],
                        part_state_path=str(rec_cache.base_dir_path.absolute()),
                        skip_check_same_file=self.conf.skip_check_same_file,
                    )
                    all_completed = all_completed and _completed

                    _uploaded_files.append(_abs_path)
                    rec_cache.uploaded_filepaths = _uploaded_files
                    rec_cache.save_state()

                    task_name = rec_cache.task.get("name", "")
                    if task_name:
                        self.api.put_task_tags(task_name, {"uploadedFiles": str(len(_uploaded_files))})

                # 6. 完成记录。如果需要，删除 record 文件夹
                if all_completed:
                    _log.info("==> All files uploaded")

                    if not self._upload_finish_flag_file(rec_cache.record["name"], rec_cache):
                        _log.error(f"==> Failed to upload finish flag file: {rec_cache.key}")
                        return

                    self.api.update_record(
                        record_name=rec_cache.record["name"],
                        labels=rec_cache.labels + ["上传完成"],
                    )
                    task_name = rec_cache.task.get("name", "")
                    if task_name:
                        self.api.put_task_tags(
                            task_name, {"recordName": rec_cache.record.get("name"), "totalFiles": str(rec_cache.total_files)}
                        )
                        self.api.update_task_state(task_name, "SUCCEEDED")
                    if rec_cache.diagnosis_task.get("name", ""):
                        self.api.update_task_state(rec_cache.diagnosis_task.get("name"), "SUCCEEDED")

                    rec_cache.uploaded = True
                    rec_cache.save_state()
                    _log.info(f"==> Handled record: {rec_cache.key}")

                    if self.conf.delete_after_upload:
                        rec_cache.delete_cache_dir(delay_in_hours=0)

    # noinspection PyBroadException
    def run(self, network_queue: Queue, error_queue: Queue):
        _log.info(f"==> Search for new record in {RECORD_DIR_PATH}")
        total_records = 0
        for record in RecordCache.find_all():
            try:
                self.handle_record(record)
                total_records += 1
            except Unauthorized:
                _log.error(f"==> Unauthorized when handling: {record.key}", exc_info=True)
                state = ApiClientState().load_state()
                state.authorized_device(0, "")
                state.save_state()
            except RecordNotFound:
                _log.error(f"==> Record already been deleted: {record.key}, skip handle!")
                record.delete_cache_dir(delay_in_hours=0)
            except Exception as e:
                # 打印错误，但保证循环不被打断
                _log.error(f"An error occurred when handling: {record.key}", exc_info=True)
                error_queue.put({"code": type(e).__name__, "error_msg": str(e)})

            # 不管上述结果如何，超过一定时间后，删除 record 文件夹
            _log.debug(f"==> Record previously uploaded: {record.key}")
            record.delete_cache_dir(self.conf.delete_after_interval_in_hours)

            try:
                network_queue.put(
                    {"upload": request_hook.get_network_upload_usage(), "download": request_hook.get_network_download_usage()}
                )
                request_hook.reset_network_usage()
            except Exception:
                _log.error("update network usage error", exc_info=True)

File: internal/collector/collector.go
// Copyright 2025 coScene
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package collector

import (
	"context"
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	openAnaV1alpha1Enum "buf.build/gen/go/coscene-io/coscene-openapi/protocolbuffers/go/coscene/openapi/analysis/v1alpha1/enums"
	openAnaV1alpha1Resource "buf.build/gen/go/coscene-io/coscene-openapi/protocolbuffers/go/coscene/openapi/analysis/v1alpha1/resources"
	openDpsV1alpha1Enum "buf.build/gen/go/coscene-io/coscene-openapi/protocolbuffers/go/coscene/openapi/dataplatform/v1alpha1/enums"
	openDpsV1alpha1Resource "buf.build/gen/go/coscene-io/coscene-openapi/protocolbuffers/go/coscene/openapi/dataplatform/v1alpha1/resources"
	"github.com/ThreeDotsLabs/watermill/pubsub/gochannel"
	"github.com/coscene-io/coscout/internal/api"
	"github.com/coscene-io/coscout/internal/config"
	"github.com/coscene-io/coscout/internal/core"
	"github.com/coscene-io/coscout/internal/model"
	"github.com/coscene-io/coscout/internal/name"
	"github.com/coscene-io/coscout/internal/storage"
	"github.com/coscene-io/coscout/pkg/constant"
	"github.com/coscene-io/coscout/pkg/utils"
	log "github.com/sirupsen/logrus"
	"google.golang.org/protobuf/types/known/durationpb"
	"google.golang.org/protobuf/types/known/structpb"
	"google.golang.org/protobuf/types/known/timestamppb"
	"gopkg.in/yaml.v3"
)

func FindAllRecordCaches() []string {
	baseFolder := config.GetRecordCacheFolder()
	var records []string
	if !utils.CheckReadPath(baseFolder) {
		return records
	}

	err := filepath.Walk(baseFolder, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			log.Errorf("walk through cache directory failed: %v", err)
			return err
		}
		if info.IsDir() {
			return nil
		}
		if strings.HasSuffix(path, ".cos/state.json") {
			records = append(records, path)
		}
		return nil
	})
	if err != nil {
		log.Errorf("walk through cache directory failed: %v", err)
	}
	return records
}

func Collect(ctx context.Context, reqClient *api.RequestClient, confManager *config.ConfManager, pubSub *gochannel.GoChannel, errorChan chan error) error {
	uploadChan := make(chan *model.RecordCache, 10)
	triggerChan := make(chan struct{}, 1)

	go Upload(ctx, reqClient, confManager, uploadChan, errorChan)
	ticker := time.NewTicker(config.CollectionInterval)
	defer ticker.Stop()

	go func(t *time.Ticker) {
		defer log.Warn("collector ticker stopped")
		for {
			select {
			case <-t.C:
				select {
				case triggerChan <- struct{}{}:
				default: // Skip if there's already a pending trigger
				}
			case <-ctx.Done():
				return
			}
		}
	}(ticker)

	go func() {
		defer log.Warn("collector goroutine stopped")
		for {
			select {
			case <-ctx.Done():
				return
			case <-triggerChan:
				appConfig := confManager.LoadWithRemote()
				getStorage := confManager.GetStorage()

				//nolint: contextcheck // context is checked in the parent goroutine
				err := handleRecordCaches(uploadChan, reqClient, appConfig, getStorage)
				if err != nil {
					errorChan <- err
				}

				time.Sleep(1 * time.Second)
			}
		}
	}()

	go triggerUpload(ctx, pubSub, triggerChan)

	<-ctx.Done()
	log.Warnf("collector context done")
	return nil
}

func triggerUpload(ctx context.Context, pubSub *gochannel.GoChannel, triggerChan chan struct{}) {
	messages, err := pubSub.Subscribe(ctx, constant.TopicCollectMsg)
	if err != nil {
		log.Errorf("subscribe to collect message failed: %v", err)
		return
	}

	for {
		select {
		case <-ctx.Done():
			return
		case msg := <-messages:
			if msg == nil {
				log.Warn("received nil collect message, skip")
				continue
			}

			log.Infof("Received collect message, start collecting for upload")
			msg.Ack()

			time.Sleep(1 * time.Second)

			select {
			case triggerChan <- struct{}{}:
			default: // Skip if there's already a pending trigger
			}
		}
	}
}

func handleRecordCaches(uploadChan chan *model.RecordCache, reqClient *api.RequestClient, config *config.AppConfig, storage *storage.Storage) error {
	log.Infof("Start collecting record caches")

	deviceInfo := core.GetDeviceInfo(storage)
	if deviceInfo == nil || deviceInfo.GetName() == "" {
		log.Warn("device info not found, skip collecting")
		//nolint: err113 // we need to check if device info is found
		return errors.New("device info not found")
	}

	// get all record caches
	records := FindAllRecordCaches()
	if len(records) == 0 {
		log.Info("No record cache found, skip collecting")
		return nil
	}

	// upload all record caches
	for _, record := range records {
		log.Infof("Processing record cache: %s", record)

		if !utils.CheckReadPath(record) {
			log.Warnf("record cache %s not exist or has no permission", record)
			continue
		}

		data, err := os.ReadFile(record)
		if err != nil {
			log.Errorf("read record cache failed: %v", err)
			continue
		}

		rc := model.RecordCache{}
		if err := json.Unmarshal(data, &rc); err != nil {
			log.Errorf("unmarshal record cache failed: %v", err)
			continue
		}

		// check if record cache is expired
		if checkRecordCacheExpired(config.Collector.DeleteAfterIntervalInHours, rc.Timestamp, record) {
			log.Infof("Record cache %s is expired, delete it", record)
			continue
		}

		if rc.Uploaded || rc.Skipped {
			log.Infof("Record cache %s has been uploaded or skipped, skip it", record)
			continue
		}

		// create related resources
		createRelatedRecordResources(deviceInfo, &rc, reqClient, config.Device)

		if rcName, ok := rc.Record["name"].(string); ok && rcName != "" {
			uploadChan <- &rc
		}
	}
	log.Infof("Finish collecting record caches")
	return nil
}

func obtainEventFromMoment(deviceInfo *openDpsV1alpha1Resource.Device, rc *model.RecordCache, moment *model.Moment, recordName, recordTitle string, reqClient *api.RequestClient) {
	if moment.Name == "" {
		// displayname: moment.Title->recordTitle
		// description: moment.Description->recordTitle
		displayName := utils.GetStringOrDefault(recordTitle, moment.Title)
		description := utils.GetStringOrDefault(recordTitle, moment.Description)

		sec, nanos := utils.NormalizeFloatTimestamp(moment.Timestamp)
		moment.Timestamp = float64(sec) + float64(nanos)/1e9

		event := openDpsV1alpha1Resource.Event{
			Record:           recordName,
			DisplayName:      displayName,
			Description:      description,
			CustomizedFields: moment.Metadata,
			TriggerTime: &timestamppb.Timestamp{
				Seconds: int64(moment.Timestamp),
				// timestamp is in seconds, so we need to convert nanoseconds to seconds
				Nanos: int32((moment.Timestamp - float64(int64(moment.Timestamp))) * 1e9),
			},
			Duration: &durationpb.Duration{
				Seconds: int64(moment.Duration),
				// duration is in seconds, so we need to convert nanoseconds to seconds
				Nanos: int32((moment.Duration - float64(int64(moment.Duration))) * 1e9),
			},
			Device: deviceInfo,
		}
		if moment.RuleName != "" {
			event.Rule = &openDpsV1alpha1Resource.DiagnosisRule{
				Name: moment.RuleName,
			}
		}
		obtainEvent, err := reqClient.ObtainEvent(rc.ProjectName, &event)
		if err != nil {
			log.Errorf("obtain event failed: %v", err)
		} else {
			moment.Name = obtainEvent.GetName()
			moment.IsNew = true
		}

		err = rc.Save()
		if err != nil {
			log.Errorf("save record cache failed: %v", err)
		}
	}
}

func triggerDeviceEventFromMoment(deviceInfo *openDpsV1alpha1Resource.Device, rc *model.RecordCache, moment *model.Moment, recordName string, reqClient *api.RequestClient, deviceConfig *config.DeviceConfig) {
	if moment.Name != "" && moment.Event == nil {
		diagnosisRuleId := ""
		if diagnosisRuleName, err := name.NewDiagnosisRule(moment.RuleName); err == nil {
			diagnosisRuleId = diagnosisRuleName.Id
		}

		deviceEvent := openAnaV1alpha1Resource.DeviceEvent{
			Code:       moment.Code,
			Parameters: moment.Metadata,
			TriggerTime: &timestamppb.Timestamp{
				Seconds: int64(moment.Timestamp),
				Nanos:   int32((moment.Timestamp - float64(int64(moment.Timestamp))) * 1e9),
			},
			Duration: &durationpb.Duration{
				Seconds: int64(moment.Duration),
				Nanos:   int32((moment.Duration - float64(int64(moment.Duration))) * 1e9),
			},
			EventSource:     openAnaV1alpha1Enum.EventSourceEnum_DEVICE,
			Moment:          moment.Name,
			Device:          deviceInfo.GetName(),
			Record:          recordName,
			DiagnosisRuleId: diagnosisRuleId,
			DeviceContext:   getDeviceExtraInfos(deviceConfig.ExtraFiles),
		}

		err := reqClient.TriggerDeviceEvent(rc.ProjectName, &deviceEvent)
		if err != nil {
			log.Errorf("trigger device event failed: %v", err)
		}
	}
}

func upsertTaskFromMoment(rc *model.RecordCache, moment *model.Moment, recordTitle string, reqClient *api.RequestClient) {
	if !moment.Task.ShouldCreate {
		return
	}

	//nolint: nestif // we need to check if the task is new
	if moment.Name != "" && moment.Task.Name == "" {
		// displayname: moment.Task.Title->moment.Title->recordTitle
		// description: moment.Task.Description->moment.Description->recordTitle
		displayName := utils.GetStringOrDefault(recordTitle, moment.Task.Title, moment.Title)
		description := utils.GetStringOrDefault(recordTitle, moment.Task.Description, moment.Description)

		// Create new task
		task := &openDpsV1alpha1Resource.Task{
			Title:       displayName,
			Description: description,
			Assignee:    moment.Task.Assignee,
			Category:    openDpsV1alpha1Enum.TaskCategoryEnum_COMMON,
			State:       openDpsV1alpha1Enum.TaskStateEnum_PENDING,
			Detail: &openDpsV1alpha1Resource.Task_CommonTaskDetail{
				CommonTaskDetail: &openDpsV1alpha1Resource.CommonTaskDetail{
					Related: &openDpsV1alpha1Resource.CommonTaskDetail_Event{
						Event: moment.Name,
					},
				},
			},
		}
		task.GetCommonTaskDetail().SetEvent(moment.Name)

		upsertedTask, err := reqClient.UpsertTask(rc.ProjectName, task)
		if err != nil {
			log.Errorf("create task failed: %v", err)
		} else {
			// Update moment's task name with created task name
			moment.Task.Name = upsertedTask.GetName()
			err = rc.Save()
			if err != nil {
				log.Errorf("save record cache failed: %v", err)
			}
			if moment.Task.Name != "" && moment.Task.SyncTask {
				t, err := reqClient.SyncTask(rc.ProjectName, moment.Task.Name)
				if err == nil {
					log.Infof("sync task %s succeeded,response: %s", moment.Task.Name, t.GetName())
				}
			}
		}
	}
}

func createRelatedRecordResources(deviceInfo *openDpsV1alpha1Resource.Device, rc *model.RecordCache, reqClient *api.RequestClient, deviceConfig config.DeviceConfig) {
	if rc.Record["name"] == nil {
		createRecord(deviceInfo, rc, reqClient)
	}

	createRecordRelatedMoments(deviceInfo, rc, reqClient, deviceConfig)
	createRecordRelatedDiagnosisTasks(deviceInfo, rc, reqClient, deviceConfig)
	createRecordRelatedUploadTasks(deviceInfo, rc, reqClient, deviceConfig)
}

func createRecordRelatedMoments(deviceInfo *openDpsV1alpha1Resource.Device, rc *model.RecordCache, reqClient *api.RequestClient, deviceConfig config.DeviceConfig) {
	if len(rc.Moments) == 0 {
		return
	}

	recordName, ok := rc.Record["name"].(string)
	if !ok || recordName == "" {
		return
	}
	recordTitle, _ := rc.Record["title"].(string)
	rc, err := rc.Reload()
	if err != nil {
		log.Errorf("reload record cache failed: %v", err)
		return
	}

	for i := range rc.Moments {
		obtainEventFromMoment(deviceInfo, rc, &rc.Moments[i], recordName, recordTitle, reqClient)
		triggerDeviceEventFromMoment(deviceInfo, rc, &rc.Moments[i], recordName, reqClient, &deviceConfig)
		upsertTaskFromMoment(rc, &rc.Moments[i], recordTitle, reqClient)
	}
}

func createRecordRelatedDiagnosisTasks(deviceInfo *openDpsV1alpha1Resource.Device, rc *model.RecordCache, reqClient *api.RequestClient, _ config.DeviceConfig) {
	if rc.DiagnosisTask == nil {
		return
	}

	taskName, ok := rc.DiagnosisTask["name"].(string)
	if ok && taskName != "" {
		return
	}

	rc, err := rc.Reload()
	if err != nil {
		log.Errorf("reload record cache failed: %v", err)
		return
	}
	recordName, ok := rc.Record["name"].(string)
	if !ok || recordName == "" {
		return
	}
	recordTitle, _ := rc.Record["title"].(string)

	diaTask := openDpsV1alpha1Resource.Task{
		Title:       recordTitle,
		Description: getRecordDescription(recordTitle, rc),
		Category:    openDpsV1alpha1Enum.TaskCategoryEnum_DIAGNOSIS,
		State:       openDpsV1alpha1Enum.TaskStateEnum_PROCESSING,
		Tags: map[string]string{
			"recordName": recordName,
			"totalFiles": strconv.Itoa(len(rc.OriginalFiles)),
		},
	}
	diagnosisTaskDetail := openDpsV1alpha1Resource.DiagnosisTaskDetail{
		Device: deviceInfo.GetName(),
	}

	ruleName, ok := rc.DiagnosisTask["rule_name"].(string)
	if ok && ruleName != "" {
		diagnosisTaskDetail.DiagnosisRule = ruleName
	}

	ruleDisplayName, ok := rc.DiagnosisTask["rule_display_name"].(string)
	if ok && ruleDisplayName != "" {
		diagnosisTaskDetail.DisplayName = ruleDisplayName
	}

	startTime, ok := rc.DiagnosisTask["start_time"].(float64)
	if ok {
		sec, nsec := utils.NormalizeFloatTimestamp(startTime)
		diagnosisTaskDetail.StartTime = &timestamppb.Timestamp{
			Seconds: sec,
			Nanos:   nsec,
		}
	}
	endTime, ok := rc.DiagnosisTask["end_time"].(float64)
	if ok {
		sec, nsec := utils.NormalizeFloatTimestamp(endTime)
		diagnosisTaskDetail.EndTime = &timestamppb.Timestamp{
			Seconds: sec,
			Nanos:   nsec,
		}
	}
	triggerTime, ok := rc.DiagnosisTask["trigger_time"].(float64)
	if ok {
		sec, nsec := utils.NormalizeFloatTimestamp(triggerTime)
		diagnosisTaskDetail.TriggerTime = &timestamppb.Timestamp{
			Seconds: sec,
			Nanos:   nsec,
		}
	}

	diaTask.SetDiagnosisTaskDetail(&diagnosisTaskDetail)
	task, err := reqClient.CreateTask(rc.ProjectName, &diaTask)
	if err != nil {
		log.Errorf("create task failed: %v", err)
	} else {
		rc.DiagnosisTask["name"] = task.GetName()
		err = rc.Save()
		if err != nil {
			log.Errorf("save record cache failed: %v", err)
		}
	}
}

func createRecordRelatedUploadTasks(_ *openDpsV1alpha1Resource.Device, rc *model.RecordCache, reqClient *api.RequestClient, _ config.DeviceConfig) {
	if rc.UploadTask == nil {
		return
	}

	rc, err := rc.Reload()
	if err != nil {
		log.Errorf("reload record cache failed: %v", err)
		return
	}
	taskName, ok := rc.UploadTask["name"].(string)
	if !ok || taskName == "" {
		return
	}

	recordName, ok := rc.Record["name"].(string)
	if !ok || recordName == "" {
		return
	}

	tags := map[string]string{}
	tags["recordName"] = recordName

	_, err = reqClient.AddTaskTags(taskName, tags)
	if err != nil {
		log.Errorf("add task tags failed: %v", err)
	}
}

func createRecord(deviceInfo *openDpsV1alpha1Resource.Device, recordCache *model.RecordCache, reqClient *api.RequestClient) {
	if recordCache.Record["name"] != nil {
		return
	}

	title := getRecordTitle(recordCache)
	description := getRecordDescription(title, recordCache)

	labels := make([]*openDpsV1alpha1Resource.Label, 0)
	for _, label := range recordCache.Labels {
		labels = append(labels, &openDpsV1alpha1Resource.Label{
			DisplayName: label,
		})
	}

	metadata := make(map[string]string)
	t := time.Now()
	zone, offset := t.Zone()
	metadata["deviceTimezoneOffset"] = strconv.Itoa(offset)
	metadata["deviceTimezone"] = zone

	record := &openDpsV1alpha1Resource.Record{
		Title:       title,
		Description: description,
		Labels:      labels,
		Device:      deviceInfo,
		Metadata:    metadata,
	}
	ruleName, ok := recordCache.DiagnosisTask["rule_name"].(string)
	if ok {
		record.Rules = []*openDpsV1alpha1Resource.DiagnosisRule{
			{
				Name: ruleName,
			},
		}
	}

	record, err := reqClient.CreateRecord(recordCache.ProjectName, record)
	if err != nil {
		log.Errorf("create record failed: %v", err)
	}
	recordCache.Record = map[string]interface{}{
		"name":        record.GetName(),
		"title":       record.GetTitle(),
		"description": record.GetDescription(),
	}

	err = recordCache.Save()
	if err != nil {
		log.Errorf("save record cache failed: %v", err)
	}
}

func getRecordTitle(rc *model.RecordCache) string {
	title, ok := rc.Record["title"].(string)
	if ok && title != "" {
		return title
	}

	taskTitle, ok := rc.UploadTask["title"].(string)
	if ok && taskTitle != "" {
		return taskTitle
	}

	triggerTime := time.Unix(rc.Timestamp/1000, 0).Format(time.RFC3339)
	return "Record at " + triggerTime
}

func getRecordDescription(title string, rc *model.RecordCache) string {
	desc, ok := rc.Record["description"].(string)
	if ok && desc != "" {
		return desc
	}

	return title
}

func checkRecordCacheExpired(expiredHours int, timestamp int64, rcPath string) bool {
	if expiredHours < 0 {
		return false
	}

	fileTime := time.UnixMilli(timestamp)
	if time.Since(fileTime).Hours() > float64(expiredHours) {
		parentFolder := utils.GetParentFolder(utils.GetParentFolder(rcPath))
		return utils.DeleteDir(parentFolder)
	}
	return false
}

func getDeviceExtraInfos(extraFiles []string) *structpb.Value {
	extraInfoMap := make(map[string]interface{})

	for _, file := range extraFiles {
		if !utils.CheckReadPath(file) {
			log.Warnf("extra file %s not exist or has no permission", file)
			continue
		}

		if !strings.HasSuffix(file, ".yaml") && !strings.HasSuffix(file, ".yml") {
			continue
		}

		data, err := os.ReadFile(file)
		if err != nil {
			log.Errorf("read extra file %s failed: %v", file, err)
			continue
		}

		var yamlData map[string]interface{}
		if err := yaml.Unmarshal(data, &yamlData); err != nil {
			log.Errorf("unmarshal yaml file %s failed: %v", file, err)
			continue
		}

		for k, v := range yamlData {
			extraInfoMap[k] = v
		}
	}

	value, err := structpb.NewStruct(extraInfoMap)
	if err != nil {
		log.Errorf("convert extra info to structpb failed: %v", err)
		return nil
	}

	return structpb.NewStructValue(value)
}
